// @hyext/hy-ui/dist/share/components/Stepper/index.js
function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

import React, { Component } from 'react';
import { View, StyleSheet, TouchableOpacity, TextInput } from 'react-native';
import variables from '../../common/styles/variables';
import styles from './styles';
var stepperStyles = StyleSheet.create(styles);
export var Stepper =
/*#__PURE__*/
function (_Component) {
  _inheritsLoose(Stepper, _Component);

  function Stepper(props) {
    var _this;

    _this = _Component.call(this, props) || this;
    _this.textInputValue = '';

    _this.onDecrease = function () {
      var _this$props = _this.props,
          value = _this$props.value,
          step = _this$props.step,
          min = _this$props.min;
      var newValue = _this.isEmpty(value) ? min + step : value;

      _this.changeValue(newValue, -step, 'decrease');
    };

    _this.onIncrease = function () {
      var _this$props2 = _this.props,
          value = _this$props2.value,
          step = _this$props2.step,
          min = _this$props2.min;
      var newValue = _this.isEmpty(value) ? min - 1 : value;

      _this.changeValue(newValue, step, 'increase');
    };

    _this.onChangeText = function (value) {
      var newValue;

      if (!value) {
        newValue = '';
      } else {
        newValue = Number(value);

        if (isNaN(newValue)) {
          newValue = '';
        }
      }

      _this.changeValue(newValue, _this.props.step, 'input');
    };

    _this.changeValue = function (value, step, action) {
      if (step === void 0) {
        step = 1;
      }

      var _this$props3 = _this.props,
          max = _this$props3.max,
          min = _this$props3.min;
      var newValue;

      if (value === '') {
        newValue = '';
      } else {
        if (action === 'input') {
          newValue = value;
        } else {
          newValue = value + step;
        }

        if (newValue > max) {
          newValue = max;
        }

        if (newValue < min) {
          newValue = min;
        }
      }

      _this.props.onChange && _this.props.onChange(newValue, _this.props.value, action);
    };

    return _this;
  }

  var _proto = Stepper.prototype;

  _proto.isEmpty = function isEmpty(value) {
    return value == null || value === '';
  };

  _proto.render = function render() {
    var _this$props4 = this.props,
        value = _this$props4.value,
        editable = _this$props4.editable,
        style = _this$props4.style,
        operatorStyle = _this$props4.operatorStyle,
        inputStyle = _this$props4.inputStyle,
        operatorIconColor = _this$props4.operatorIconColor,
        max = _this$props4.max,
        min = _this$props4.min;
    var increasable;
    var decreasable;

    if (this.isEmpty(value)) {
      increasable = true;
      decreasable = true;
    } else {
      value = Number(value);

      if (isNaN(value)) {
        value = 0;
      }

      increasable = Boolean(value < max);
      decreasable = Boolean(value > min);
    }

    var textInputValue = this.isEmpty(value) ? '' : String(value);

    if (textInputValue === this.textInputValue) {
      textInputValue += ' ';
      this.textInputValue = textInputValue;
    } else {
      this.textInputValue = textInputValue;
    }

    return React.createElement(View, {
      style: [stepperStyles.container, style]
    }, React.createElement(TouchableOpacity, {
      activeOpacity: variables.hyOpacity,
      onPress: this.onDecrease,
      disabled: !decreasable
    }, React.createElement(View, {
      style: [stepperStyles.ctrl, operatorStyle, !decreasable ? stepperStyles.disabled : null]
    }, React.createElement(View, {
      style: [stepperStyles.ctrlSymbolHor, {
        backgroundColor: operatorIconColor
      }]
    }))), React.createElement(TextInput, {
      style: [stepperStyles.input, inputStyle],
      value: textInputValue,
      onChangeText: this.onChangeText,
      editable: editable,
      keyboardType: "numeric"
    }), React.createElement(TouchableOpacity, {
      activeOpacity: variables.hyOpacity,
      onPress: this.onIncrease,
      disabled: !increasable
    }, React.createElement(View, {
      style: [stepperStyles.ctrl, operatorStyle, !increasable ? stepperStyles.disabled : null]
    }, React.createElement(View, {
      style: [stepperStyles.ctrlSymbolHor, {
        backgroundColor: operatorIconColor
      }]
    }), React.createElement(View, {
      style: [stepperStyles.ctrlSymboVer, {
        backgroundColor: operatorIconColor
      }]
    }))));
  };

  return Stepper;
}(Component);
Stepper.defaultProps = {
  operatorIconColor: variables.hyGrayDarker,
  min: 1,
  max: 5,
  step: 1,
  editable: false
};
