// @hyext/hy-ui/dist/share/components/Stepper/styles.js
import variables from '../../common/styles/variables';
import { StyleSheet } from 'react-native';
export default {
  container: {
    flexDirection: 'row',
    width: 136,
    alignItems: 'center'
  },
  ctrl: {
    width: 30,
    height: 30,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: variables.hyBrandPrimary,
    borderRadius: variables.hyRadiusXS
  },
  squareSolid: {
    borderRadius: 4
  },
  circularHollow: {
    backgroundColor: 'transparent',
    borderWidth: StyleSheet.hairlineWidth,
    borderColor: variables.hyBrandPrimary
  },
  squareHollow: {
    borderRadius: 4,
    backgroundColor: 'transparent',
    borderWidth: StyleSheet.hairlineWidth,
    borderColor: variables.hyBrandPrimary
  },
  squareJoinHollowLeft: {
    borderTopLeftRadius: 4,
    borderBottomLeftRadius: 4,
    borderTopRightRadius: 0,
    borderBottomRightRadius: 0,
    backgroundColor: 'transparent',
    borderWidth: StyleSheet.hairlineWidth,
    borderColor: variables.hyBrandPrimary
  },
  squareJoinHollowRight: {
    borderTopLeftRadius: 0,
    borderBottomLeftRadius: 0,
    borderTopRightRadius: 4,
    borderBottomRightRadius: 4,
    backgroundColor: 'transparent',
    borderWidth: StyleSheet.hairlineWidth,
    borderColor: variables.hyBrandPrimary
  },
  ctrlSymbolHor: {
    width: 14,
    height: 2,
    backgroundColor: '#fff',
    borderRadius: 2
  },
  ctrlSymboVer: {
    position: 'absolute',
    width: 2,
    height: 14,
    backgroundColor: '#fff',
    borderRadius: 2
  },
  ctrlSymbolSolid: {
    backgroundColor: variables.hyBrandPrimary
  },
  ctrlText: {
    alignItems: 'center',
    justifyContent: 'center',
    fontSize: 28,
    color: '#fff',
    borderRadius: 100
  },
  input: {
    width: 60,
    marginHorizontal: 10,
    textAlign: 'center',
    fontSize: 14,
    color: variables.hyGrayBase,
    textAlignVertical: 'center'
  },
  inputWithBorder: {
    borderTopColor: variables.hyBrandPrimary,
    borderTopWidth: StyleSheet.hairlineWidth,
    borderBottomColor: variables.hyBrandPrimary,
    borderBottomWidth: StyleSheet.hairlineWidth
  },
  disabled: {
    opacity: variables.hyOpacity
  }
};
