// @hyext/hy-ui/dist/share/components/Longlist/index.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

import { FlatList, Text, StyleSheet, View, ActivityIndicator } from 'react-native';
import React from 'react';
import variables from '../../common/styles/variables';
var styles = StyleSheet.create({
  loadingIndicator: {
    padding: 10
  }
});
export var Longlist =
/*#__PURE__*/
function (_React$Component) {
  _inheritsLoose(Longlist, _React$Component);

  // 通过 flatList 对象，调用 FlatList 组件相关方法
  function Longlist(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;
    _this.flatList = null;

    _this.handleEndReached = function () {
      var _this$props = _this.props,
          data = _this$props.data,
          total = _this$props.total,
          onEndReached = _this$props.onEndReached;

      if (!onEndReached) {
        return;
      }

      if (data && data.length && data.length >= total) {
        return;
      }

      if (_this.state.loading) {
        return;
      }

      _this.setState({
        loading: true
      }, function () {
        onEndReached().then(function () {
          _this.setState({
            loading: false
          });
        })["catch"](function (e) {
          _this.setState({
            loading: false
          });
        });
      });
    };

    _this.handleRefresh = function () {
      if (_this.state.refreshing) {
        return;
      }

      _this.setState({
        refreshing: true
      }, function () {
        _this.props.onRefresh().then(function () {
          _this.setState({
            refreshing: false
          });
        })["catch"](function () {
          _this.setState({
            refreshing: false
          });
        });
      });
    };

    _this.state = {
      refreshing: false,
      loading: false
    };
    return _this;
  }

  var _proto = Longlist.prototype;

  _proto.renderFooter = function renderFooter() {
    var _this$props2 = this.props,
        data = _this$props2.data,
        total = _this$props2.total,
        renderFooter = _this$props2.renderFooter;
    var loading = this.state.loading;
    var footer = null;

    if (renderFooter) {
      footer = renderFooter(loading, data, total);
    }

    if (React.isValidElement(footer)) {
      return footer;
    }

    if (loading) {
      return React.createElement(View, {
        style: styles.loadingIndicator
      }, React.createElement(ActivityIndicator, {
        size: "small",
        color: "#333"
      }));
    }

    if (data && !data.length && total === 0) {
      return React.createElement(Text, {
        style: {
          padding: variables.hyHSpacingXL,
          color: variables.hyGrayBase,
          textAlign: 'center'
        }
      }, "\u65E0\u6570\u636E");
    }

    if (data && data.length && data.length >= total) {
      return React.createElement(Text, {
        style: {
          padding: variables.hyHSpacingXL,
          color: variables.hyGrayBase,
          textAlign: 'center'
        }
      }, "\u65E0\u66F4\u591A\u6570\u636E");
    }

    return null;
  };

  _proto.render = function render() {
    var _this2 = this;

    var refreshing = this.state.refreshing;
    var onRefresh = this.props.onRefresh;

    var retProps = _objectSpread({}, this.props);

    if (!onRefresh) {
      delete retProps.refreshing;
      delete retProps.onRefresh;
    } else {
      retProps.refreshing = refreshing;
      retProps.onRefresh = this.handleRefresh;
    }

    return React.createElement(FlatList, _extends({}, retProps, {
      ref: function ref(c) {
        _this2.flatList = c;
      },
      keyExtractor: function keyExtractor(item, index) {
        return index.toString();
      },
      initialNumToRender: this.props.initialNumToRender,
      onEndReached: this.handleEndReached,
      onEndReachedThreshold: 0.1,
      ListFooterComponent: function ListFooterComponent() {
        return _this2.renderFooter();
      },
      onScrollToIndexFailed: function onScrollToIndexFailed() {}
    }));
  };

  return Longlist;
}(React.Component);
Longlist.defaultProps = {
  total: 0,
  data: [],
  initialNumToRender: 5
};
