// @hyext/hy-ui/dist/share/components/Form/styles.js
import variables from '../../common/styles/variables';
import { StyleSheet } from 'react-native';
export default {
  form: {
    backgroundColor: '#fff'
  },
  title: {
    paddingHorizontal: variables.hyHSpacingL,
    paddingTop: variables.hyVSpacingL,
    paddingBottom: variables.hyVSpacingL - 5,
    fontSize: variables.hyFontSizeM,
    color: variables.hyGrayDarker
  },
  label: {
    minWidth: variables.formItemLabelWidth,
    marginRight: variables.formItemLabelMarginRight,
    flexDirection: 'row',
    alignItems: 'center'
  },
  labelText: {
    fontSize: variables.hyFontSizeM,
    color: variables.hyGrayBase
  },
  errorText: {
    fontSize: variables.hyFontSizeM,
    color: variables.hyBrandDanger
  },
  indicator: {
    paddingLeft: variables.hyHSpacingL
  },
  line: {
    marginLeft: variables.hyHSpacingL,
    height: StyleSheet.hairlineWidth,
    backgroundColor: variables.hyBorderColorDark
  },
  formItem: {
    paddingHorizontal: variables.formItemHSpacing,
    paddingVertical: variables.formItemVSpacing
  },
  container: {
    flexDirection: 'row',
    alignItems: 'center'
  },
  control: {
    flex: 1
  },
  mask: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    backgroundColor: 'rgba(255, 255, 255, 0.5)'
  },
  validationView: {
    color: variables.hyBrandDanger,
    fontSize: variables.hyFontSizeS
  }
};
