// @hyext/hy-ui/dist/share/components/Tag/styles.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import { StyleSheet } from 'react-native';
import variables from '../../common/styles/variables';
var defaultWrapper = {
  flexDirection: 'row',
  justifyContent: 'center',
  alignItems: 'center',
  paddingHorizontal: variables.hyHSpacingM,
  paddingVertical: variables.hyVSpacingS,
  borderWidth: StyleSheet.hairlineWidth,
  borderColor: variables.hyGray,
  borderRadius: variables.hyRadiusXS,
  backgroundColor: variables.hyFillBody
};
var defaultText = {
  fontSize: variables.hyFontSizeS,
  color: variables.hyGrayDarker
};

var primaryWrapper = _objectSpread({}, defaultWrapper, {
  borderColor: variables.hyBrandPrimary,
  backgroundColor: variables.hyBrandPrimary
});

var primaryText = _objectSpread({}, defaultText, {
  color: '#fff'
});

var dangerWrapper = _objectSpread({}, defaultWrapper, {
  borderColor: variables.hyBrandDanger,
  backgroundColor: variables.hyBrandDanger
});

var dangerText = _objectSpread({}, defaultText, {
  color: '#fff'
});

var warningWrapper = _objectSpread({}, defaultWrapper, {
  borderColor: variables.hyBrandWarning,
  backgroundColor: variables.hyBrandWarning
});

var warningText = _objectSpread({}, defaultText, {
  color: '#fff'
});

var successWrapper = _objectSpread({}, defaultWrapper, {
  borderColor: variables.hyBrandSuccess,
  backgroundColor: variables.hyBrandSuccess
});

var successText = _objectSpread({}, defaultText, {
  color: '#fff'
});

var infoWrapper = _objectSpread({}, defaultWrapper, {
  borderColor: variables.hyBrandInfo,
  backgroundColor: variables.hyBrandInfo
});

var infoText = _objectSpread({}, defaultText, {
  color: '#fff'
});

export default StyleSheet.create({
  defaultWrapper: defaultWrapper,
  defaultText: defaultText,
  primaryWrapper: primaryWrapper,
  primaryText: primaryText,
  dangerWrapper: dangerWrapper,
  dangerText: dangerText,
  successWrapper: successWrapper,
  successText: successText,
  warningWrapper: warningWrapper,
  warningText: warningText,
  infoWrapper: infoWrapper,
  infoText: infoText
});
