// @hyext/hy-ui/dist/share/common/styles/variables.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import { StyleSheet, Platform } from 'react-native'; // 品牌色

var hyBrandColors = {
  hyBrandPrimary: '#ffd800',
  hyBrandPrimaryDark: '#FFA000',
  hyBrandSuccess: '#61cb28',
  hyBrandWarning: '#ff8400',
  hyBrandDanger: '#f23244',
  hyBrandInfo: '#188afa'
}; // 灰度

var hyGrayColors = {
  hyGrayBase: '#111',
  // 正文，主标题
  hyGrayDarker: '#333',
  // 副标题
  hyGrayDark: '#555',
  // 补充、提示信息
  hyGray: '#888',
  // 取消按钮等
  hyGrayLight: '#aaaaaa',
  hyGrayLighter: '#cccccc',
  hyGrayLightest: '#ebebeb'
}; // 背景色

var hyFillColors = {
  hyFillBase: '#ffffff',
  hyFillGray: '#F5F5F5',
  hyFillBody: '#F8F8F8',
  hyFillBackdrop: 'rgba(0, 0, 0, .3)',
  hyFillBackdropDark: 'rgba(0, 0, 0, 0.75)'
}; // 字体尺寸

var hyFontSize = {
  hyFontSizeXS: 10,
  hyFontSizeS: 12,
  hyFontSizeM: 14,
  hyFontSizeL: 16,
  hyFontSizeXL: 18,
  hyFontSizeX2L: 20,
  hyFontSizeX3L: 22,
  hyFontSizeX4L: 24,
  hyFontSizeX5L: 28
};
var hySpacing = {
  // 水平间距
  hyHSpacingS: 4,
  hyHSpacingM: 8,
  hyHSpacingL: 12,
  hyHSpacingXL: 16,
  hyHSpacingX2L: 20,
  // 垂直间距
  hyVSpacingXS: 2,
  hyVSpacingS: 4,
  hyVSpacingM: 8,
  hyVSpacingL: 10,
  hyVSpacingXL: 12,
  hyVSpacingX2L: 16,
  hyVSpacingX3L: 18,
  hyVSpacingX4L: 20
}; // 圆角

var hyRadius = {
  hyRadiusXS: 2,
  hyRadiusS: 4,
  hyRadiusM: 6,
  hyRadiusL: 8
};
var hyBorder = {
  hyBorderWidth: StyleSheet.hairlineWidth,
  hyBorderColor: '#F5F5F5',
  hyBorderColorDark: '#e5e5e5',
  hyBorderColorDarker: '#d5d5d5'
};
var hyOpacity = 0.3;
var hyEnableAnimated = true;
/**
 * Button 组件
 */

var button = {
  buttonEnableAnimated: Platform.OS === 'ios',
  buttonBorderRadius: hyRadius.hyRadiusXS,
  buttonActiveOpacity: hyOpacity,
  buttonLFontSize: hyFontSize.hyFontSizeXL,
  buttonLHSpacing: 50,
  buttonLVSpacing: 14,
  buttonMFontSize: hyFontSize.hyFontSizeL,
  buttonMHSpacing: 46,
  buttonMVSpacing: 12,
  buttonSFontSize: hyFontSize.hyFontSizeM,
  buttonSHSpacing: hySpacing.hyHSpacingXL,
  buttonSVSpacing: 8
};
/**
 * Form 组件
 */

var form = {};
var formItem = {
  formItemHSpacing: hySpacing.hyHSpacingXL,
  formItemVSpacing: 18,
  formItemLabelWidth: 90,
  formItemLabelMarginRight: 32
};
/**
 * Input组件
 */

var input = {
  // input组件安全区域 高度
  inputTextFontSize: 14,
  inputAreaHeight: 30
};
var radio = {
  radioEnableAnimated: hyEnableAnimated
};
var checkbox = {
  checkboxEnableAnimated: hyEnableAnimated
};
var topview = {
  topviewZIndex: 100
};
var slider = {};
var dropdown = {
  dropdownEnableAnimated: hyEnableAnimated
};

var variables = _objectSpread({}, hyBrandColors, {}, hyGrayColors, {}, hyFillColors, {}, hyFontSize, {}, hySpacing, {}, hyRadius, {}, hyBorder, {
  hyOpacity: hyOpacity,
  hyEnableAnimated: hyEnableAnimated
}, button, {}, form, {}, formItem, {}, input, {}, radio, {}, checkbox, {}, slider, {}, topview, {}, dropdown);

function useTheme(args) {
  if (args === void 0) {
    args = {};
  }

  Object.assign(variables, args);
  return variables;
}

export default variables;
export { useTheme };
